src/orca-jedi/nemo_io/AtlasIndex.h

```diff
@@ -0,0 +1,214 @@
+#include "atlas-orca/grid/OrcaGrid.h"
+
+namespace orcamodel {
+
+/// \brief Interface from netcdf i, j coordinates in a field to 1D index of atlas
+///        field data.
+class AtlasIndexToBufferIndex {
+ public:
+  virtual ~AtlasIndexToBufferIndex() {}
+  virtual int64_t operator()(const int i, const int j) const = 0;
+  virtual int64_t operator()(const size_t inode) const = 0;
+  virtual std::pair<int, int> ij(const size_t inode) const  = 0;
+  virtual size_t nx() const = 0;
+  virtual size_t ny() const = 0;
+};
+
+/// \brief Indexer from the orca netcdf i, j field to an index of a 1D buffer
+///        of orca data.
+class OrcaIndexToBufferIndex : public AtlasIndexToBufferIndex {
+ private:
+  atlas::OrcaGrid orcaGrid_;
+  int32_t ix_glb_max;
+  int32_t iy_glb_max;
+  int32_t glbarray_offset;
+  int32_t glbarray_jstride;
+  size_t nx_;
+  size_t ny_;
+  const atlas::Mesh mesh_;
+
+ public:
+  static std::string name() {return "ORCA";}
+  size_t nx() const {return nx_;}
+  size_t ny() const {return ny_;}
+
+  explicit OrcaIndexToBufferIndex(const atlas::Mesh& mesh) : orcaGrid_(mesh.grid()), mesh_(mesh) {
+    iy_glb_max = orcaGrid_.ny() + orcaGrid_.haloNorth() - 1;
+    ix_glb_max = orcaGrid_.nx() + orcaGrid_.haloEast() - 1;
+
+    nx_ = orcaGrid_.nx() + orcaGrid_.haloWest() + orcaGrid_.haloEast();
+    ny_ = orcaGrid_.ny() + orcaGrid_.haloSouth() + orcaGrid_.haloNorth();
+
+    // vector of local indices: necessary for remote indices of ghost nodes
+    int iy_glb_min = -orcaGrid_.haloSouth();
+    int ix_glb_min = -orcaGrid_.haloWest();
+    glbarray_offset  = -(nx_ * iy_glb_min) - ix_glb_min;
+    glbarray_jstride = nx_;
+  }
+
+  /// \brief Index of a 1D array corresponding to point i, j
+  /// \param i
+  /// \param j
+  /// \return index of a matching 1D array
+  int64_t operator()(const int i, const int j) const {
+      ATLAS_ASSERT(i <= ix_glb_max,
+          std::to_string(i) + " > " + std::to_string(ix_glb_max));
+      ATLAS_ASSERT(j <= iy_glb_max,
+          std::to_string(j) + " > " + std::to_string(iy_glb_max));
+      return glbarray_offset + j * glbarray_jstride + i;
+  }
+
+  /// \brief Index of a 1D array corresponding to a mesh node index
+  /// \param inode
+  /// \return index of a matching 1D array
+  int64_t operator()(const size_t inode) const {
+      auto ij = atlas::array::make_view<int32_t, 2>(mesh_.nodes().field("ij"));
+      return (*this)(ij(inode, 0), ij(inode, 1));
+  }
+
+  /// \brief i, j pair corresponding to a node number. Only use this for diagnostic purposes as
+  ///        it is not robust at the orca halo.
+  /// \param inode
+  /// \return std::pair of the 2D indices corresponding to the node.
+  std::pair<int, int> ij(const size_t inode) const {
+    auto ij = atlas::array::make_view<int32_t, 2>(mesh_.nodes().field("ij"));
+    int i = ij(inode, 0) >= 0 ? ij(inode, 0) : nx_ + ij(inode, 0);
+    const int ci = i >= nx_ ? i - nx_ : i;
+    int j = ij(inode, 1) + 1;
```

Comment:
This function is just here for diagnostic purposes, I would refactor it once atlas-orca contains our changes to the orca mesh and grid. The +1 is here because there is always a southern halo, so the ij field starts at j = -1. This is no good if you want to just visualise the field in 2D. This method does scuff up the north halo, but as I say its just a diagnostic function at this point and isn't actually used anywhere.

My issue is, I want it to be part of the `AtlasIndex.h` interface, so I need to put something in here. I suppose I could remove it and replace with raising a not implemented error if you prefer?

Reply:
Thanks for the clarification. I think it is fine to leave it as it is for now in that case.